python/ray/data/aggregate.py

```diff
@@ -19,8 +28,9 @@ class AggregateFn(object):
     def __init__(
         self,
         init: Callable[[KeyType], AggType],
-        accumulate: Callable[[AggType, T], AggType],
         merge: Callable[[AggType, AggType], AggType],
+        accumulate: Callable[[AggType, T], AggType] = None,
+        accumulate_block: Callable[[AggType, Block[T]], AggType] = None,
```

Comment:
Looked at the implementation seems instead of `Block`, `BlockAccessor` is passed in?

Is BlockAccessor part of the public api, I feel we should pass in Block instead of BlockAccessor

Reply:
Reasoning for using `BlockAccessor`: using `BlockAccessor` as the argument type is much cleaner for aggregation implementers _and_ for the aggregation callers (i.e. `BlockAccessor.combine()`), since the aggregations can rely on standardized APIs across block types and the callers don't have to convert the `BlockAcessor` back to a `Block`. If we change the argument type to `Block`, `BlockAccessor.combine()` will convert a `BlockAccessor` to a `Block`, just for every aggregation to immediately convert the block to a `BlockAccessor`.

I suppose that the most common case for a user implementing an aggregation is that they'll be implementing it with a specific `Block` in mind, so them being able to define `accumulate_block` via the e.g. Pandas DataFrame API will be better UX than them needing to learn our `BlockAccessor` APIs...

I'll change it back to `Block` since the UX for users implementing custom aggregations is the most important thing.

Reply:
I have a side question: the BlockAccessor is under ray/data, not ray/data/impl; so are we relying on @PublicAPI to separate public v.s. internal? I think it's a good practice to establish a clear boundary (whichever way we choose).

Reply:
Yep `@PublicAPI` is our primary public vs. internal separator. So far, for abstractions/modules that have any public APIs, we've kept them outside of `ray/data/impl`, and then we gradually try to move the non-public bits of those abstractions/modules into `ray/data/impl` as it grows, when the separation makes sense.